rfcs/20200624-pluggable-device-for-tensorflow.md

```diff
@@ -0,0 +1,347 @@
+# Pluggable device for TensorFlow
+
+| Status        | Proposed                                             |
+:-------------- |:---------------------------------------------------- |
+| **RFC #**     | [262](https://github.com/tensorflow/community/pull/262)|
+| **Author(s)** | Zhoulong Jiang (zhoulong.jiang@intel.com), Yiqiang Li (yiqiang.li@intel.com),  Eric Lin (eric.lin@intel.com), Jianhui Li (jian.hui.li@intel.com) |
+| **Sponsor**   | Anna Revinskaya (annarev@google.com)                 |
+| **Updated**   | 2020-08-13                                           |
+
+## **Objective**
+
+Implement a pluggable device mechanism which allows to run existing TensorFlow programs on a new device without user changing the code.  Users only need to install the plugin in a specified directory, and the mechanism is able to discover and plug in the capabilities offered by the plugin. 
+
+This RFC is based on the Modular TensorFlow  [RFC](https://github.com/tensorflow/community/pull/77), which aims at extending the TensorFlow design to plug in capabilities like adding a new device support.  The modular device interface is based on StreamExecutor C API [RFC](https://github.com/tensorflow/community/pull/257). 
+
+## **Motivation**
+
+When extending TensorFlow to support a new device, one needs to modify TensorFlow code and maintain a special TensorFlow build for the new device. Modular TensorFlow RFC design a plugin architecture for serveral TensorFlow components(`Networking`, `Filesystems`, `Kernel`, `Graph` and `Accelerator backends`). This RFC describes the Accelerator backends module in the TensorFlow proper side, by introducing pluggable device to the TensorFlow device classes.
+
+The pluggable device discovery and initialization is transparent to end users. As long as the device plugin libraries follow the design described in this RFC, it can be plugged to TensorFlow proper and enable TensorFlow to run existing TensorFlow programs on a new device. 
```

Comment:
Would there be cases that explicit Python initialization API is desired? e.g., expensive initialization, initialization with custom parameters.  For example TF, has `tf.tpu.experimental.initialize_tpu_system()` API.

Reply:
Currently we didn’t find the case that it will cause expensive initialization. During the tensorflow import, it will only load the plugins and fill the C API handle. For custom parameters, I think we can have new extension to support this.